### PR TITLE
Fixed the DQN example and added Base.length to discrete

### PR DESCRIPTION
**Pull request by kraftpunk97-zz**
The DQN example for the CartPole environment and the environment itself has been updated to stop converting the actions (`Discrete(2)`) to `(-1, 1)`. Function `inv_action` has been discarded. ~Please make corresponding changes in the model-zoo as well.~ Changes proposed to the model-zoo version with [this](https://github.com/FluxML/model-zoo/pull/142) PR.

### CODE REVIEW

File: src/Envs/classic_control/CartPole.jl [hunk 1]
```diff
@@ -60,11 +60,11 @@ function CartPoleEnv()
...
 
 function step!(env::CartPoleEnv, action)
-    # @assert action ∈ env.action_space "$action in ($(env.action_space)) invalid"
+    @assert action ∈ env.action_space "$action in ($(env.action_space)) invalid"
     state = env.state
     x, ẋ, θ, θ̇  = state[1:1], state[2:2], state[3:3], state[4:4]
-    # force = action == 2 ? env.force_mag : -env.force_mag
-    force = action .* env.force_mag  # action is +1 or -1
+    # Action is either 1 or 2. Force is either -force_mag or +force_mag
+    force = (2action-3) .* env.force_mag
```
Comment (tejank10) [inline src/Envs/classic_control/CartPole.jl]: The DiffRL example gives the action output as +/- 1. We'll have to transform that action as `(3-action)/2` before passing to `step!`.

Comment (kraftpunk97-zz) [reply src/Envs/classic_control/CartPole.jl]: Ok, this turned out to be more challenging than expected😅.

Simply updating the `action()` function wasn't enough, as the program wasn't running and the error messages were not very helpful. A lot of troubleshooting later, I was finally able to isolate the problem, which was my assumption that the `action` is simply an Integer, and not a single element array. 

So fixed that, but there's a new error now, which is unavoidable. The `contains` function for the Discrete Space uses `try` and `catch` to find if the value being checked belongs to the Space. However, Zygote can not handle control flow jumps like `try` and `catch`. So the program crashes with a `Zygote.CompileError`. If the assertion check is removed, I get a very ugly Internal Error by buildbot, but the program runs, and takes ~10 episodes to train. 

How do we proceed?

Comment (kraftpunk97-zz) [reply src/Envs/classic_control/CartPole.jl]: Pinging @tejank10

Comment (tejank10) [reply src/Envs/classic_control/CartPole.jl]: We can use `@nograd` on the `contains` function, as its gradients are not required.

**Comment by kraftpunk97-zz** [comment]
@tejank10 Made the requested changes. Please review.

**Review by tejank10 — COMMENTED** [review]
Let's have `Manifest.toml` and `Project.toml` for DQN.jl.

**Comment by kraftpunk97-zz** [comment]
✅**Check**

File: examples/CartPole/DQN.jl [hunk 2]
```diff
@@ -48,14 +48,11 @@ remember(state, action, reward, next_state, done) =
   push!(memory, (data(state), action, reward, data(next_state), done))
 
 function action(state, train=true)
-  train && rand() <= get_ϵ(e) && (return rand(-1:1))
+  train && rand() <= get_ϵ(e) && (return Gym.sample(env.action_space))
```
Comment (tejank10) [inline examples/CartPole/DQN.jl:51]: `env.action_space` -> `env._env.action_space`. I think we should have a function `action_space` to access it.

Comment (kraftpunk97-zz) [reply examples/CartPole/DQN.jl:51]: I'm not so sure about that. The action space is already a part of the `EnvWrapper`. And there's nothing that you can do with the action space, except for using them to draw samples and check whether the entered action is valid or not. Adding another helper function  that just returns the action space of the environment seems unnecessary.

**Comment by kraftpunk97-zz** [comment]
@tejank10 I looked into the internal error I got with the DiffRL.jl example (Present in [this gist](https://gist.github.com/kraftpunk97/0f65971784b631cf509bae54faf4f512)) So, it all comes down to line 90 of the CartPole environment which is...

    done =  !(all(vcat(-env.x_threshold .≤ x_ .≤ env.x_threshold, -env.θ_threshold_radians .≤ θ_ .≤ env.θ_threshold_radians)))

Any clue as to why this doesn't sit right with Zygote?

File: examples/CartPole/DQN.jl (hunk 2 above)
Comment (tejank10) [reply examples/CartPole/DQN.jl:51]: Could not find `action_space` as part of `EnvWrapper` on master. Where did you find it?

Comment (kraftpunk97-zz) [reply examples/CartPole/DQN.jl:51]: https://github.com/kraftpunk97/Gym.jl/commit/b9b950e6fa2784d7526499b6e5279c01f6cbac39

**Pull request merged**